### ISSUE TITLE
[tracy] support for disabled tracy

### ISSUE THREAD

**Issue by qbojj** [microsoft/vcpkg#36725]
### Is your feature request related to a problem? Please describe.

Tracy has a `TRACY_ENABLE` flag that makes all its functions NO-OPS (empty defines). What it means, that users of the library can use the same source without additional ifdefs for release for end users and internal performance testing.

It would be helpful if vcpkg had a "null" port of tracy for use in final releases.

### Proposed solution

Add a (possibly default) feature that sets `TRACY_ENABLE` to `ON` (otherwise `OFF`)

**Comment by jimwang118** [comment]
TRACY_ENABLE: ON has been set in the [cmakelists.txt](https://github.com/wolfpld/tracy/blob/0295ef2b3cbfa91c6b9bf6bd163a5f9d61244a81/CMakeLists.txt#L65) file of the upstream code, and no additional settings are required.

**Comment by dg0yt** [comment]
Couldn't a default feature `profiling` solve the issue?

> TRACY_ENABLE: ON has been set in the [cmakelists.txt](https://github.com/wolfpld/tracy/blob/0295ef2b3cbfa91c6b9bf6bd163a5f9d61244a81/CMakeLists.txt#L65) file of the upstream code, and no additional settings are required.

Misses the point.

**Comment by maxime-modulopi** [comment]
Not being able to set `TRACY_ENABLE` to OFF makes the `tracy` port basically useless for real projects as there is no way to disable Tracy for production builds.

**Comment by andre-nguyen** [comment]
I agree that this is important, but it is not clear to me how to best fix it. Making a vcpkg feature to set `TRACY_ENABLE` has the downside that you have to reinstall a package to disable profiling; as effectively blocks using manifest mode with two build configurations (one with and one without profiling). 

My preferred solution would be to set `TRACY_ENABLE` to `OFF` like
```cmake
vcpkg_cmake_configure(
    SOURCE_PATH ${SOURCE_PATH}
    OPTIONS
        -DDOWNLOAD_CAPSTONE=OFF
        -DLEGACY=ON
        -DTRACY_ENABLE=OFF
        ${FEATURE_OPTIONS}
    OPTIONS_RELEASE
        ${TOOLS_OPTIONS}
    MAYBE_UNUSED_VARIABLES
        DOWNLOAD_CAPSTONE
        LEGACY
)
```
And then tell the user in the usage notes to define the `TRACY_ENABLE` macro themselves like
```
tracy provides CMake targets:

  find_package(Tracy REQUIRED)
  ...
  target_compile_definitions(main PRIVATE -DTRACY_ENABLE)
  target_link_libraries(main PRIVATE Tracy::TracyClient)
```

But this has the downside that many users might end up thinking the tracy package is broken.

@dg0yt @maxime-modulopi any thoughts?

**Comment by dg0yt** [comment]
This is what I see and understand:

The `TRACY_ENABLE` setting is 
- burnt into the binary lib.
- burnt into the CMake config's `INTERFACE_COMPILE_DEFINITIONS`.

The first might not be a problem because the functions could be just dead code...
...until linking to `Tracy::TracyClient` forces the header to emit actual calls into the library, due to the `INTERFACE_COMPILE_DEFINITIONS`.

I do share the concern that requiring an explicit `-DTRACY_ENABLE` might be confusing. Default behavior should be as expected.

But vcpkg port could modify the installed header to make `-DUNOFFICIAL_TRACY_DISABLE` override the `-DTRACY_ENABLE`. (Or somebody asks upstream, for coordinating an official name for such a macro.)

**Comment by andre-nguyen** [comment]
# Other solution: leave it up to the end user

It would also be possible for an end user to remove the `TRACY_ENABLE` definition by removing it from the target interface if they want to disable profiling:

```cmake
find_package(Tracy REQUIRED)

if (DISABLE_PROFILING)
  get_target_property(TRACY_INTERFACE Tracy::TracyClient INTERFACE_COMPILE_DEFINITIONS)
  list(REMOVE_ITEM TRACY_INTERFACE "TRACY_ENABLE")
  set_target_properties(Tracy::TracyClient PROPERTIES INTERFACE_COMPILE_DEFINITIONS "${TRACY_INTERFACE}")
endif ()
```

This could be added to the usage notes.

> But the vcpkg port could modify the installed header to make `-DUNOFFICIAL_TRACY_DISABLE` override the `-DTRACY_ENABLE`. (Or somebody asks upstream, for coordinating an official name for such a macro.)

Certainly possible, but it would be a large patch.

**Comment by dg0yt** [comment]
> Certainly possible, but it would be a large patch.

10 installed headers affected, one line each. Could be done with a post-install glob/replace.

> to remove the `TRACY_ENABLE` definition by removing it from the target interface

This might be tricky with subdirectories. And it is more than a single line for each user.

---

Next idea (untested): Guard the imported definition with a generator expression. One replacement after installation.
~~~diff
-   INTERFACE_COMPILE_DEFINITIONS "-DTRACY_ENABLE"
+   INTERFACE_COMPILE_DEFINITIONS "$<IF:$<BOOL:${UNOFFICIAL_TRACY_DISABLE}>,,-DTRACY_ENABLE>"
~~~~
One (optional) option in user projects.
~~~
option(UNOFFICIAL_TRACY_DISABLE "")
~~~

**Comment by rmarker** [comment]
Comment to keep the issue open.

Being able to switch the functionality on/off in different build modes is quite important.

Even if `TRACY_ENABLE` has to be added or removed manually, it would be good to document that fact.
